### ISSUE TITLE
Prohibiting depencency fetching (for test target particularly)

### ISSUE THREAD

**Issue by AMDmi3** [pypa/setuptools#1672]
My use case is packaging python modules as distro native packages (in my case, FreeBSD ports). For this case, for the sake of security, consistency and reproducibility, I need to unconditionally disallow dependency fetching (which is the case at least for `python setup.py test`) as all dependencies must be installed from the mentioned ports system, and there should be a failure if some are missing. I've found `--no-deps` argument for `easy_install` which seem to do just what I need, however I couldn't find any way to pass `easy_install` arguments via `python setup.py test`. Could it be that I'm missing something, or is this not supported? In the latter case, this should be implemented.

**Comment by pganssle** [comment]
@AMDmi3 The use of `easy_install` in general is deprecated, and we're trying to move away from anything that involves `setuptools` itself retrieving packages, in favor of `pip`.

That said, I think the safest thing to do if you *cannot* allow anything to pull in dependencies, is to manually download or build the packages you need and serve them on a local mirror of PyPI, then configure both `easy_install` and `pip` to use that as the index. It's what we're doing in the [PyPA integration tests](https://github.com/pypa/integration-test) repository. The connection to PyPI is the chokepoint that is easiest to police rather than tracking down all the possible code paths that might make that connection.

**Comment by AMDmi3** [comment]
> @AMDmi3 The use of `easy_install` in general is deprecated

We do not use it directly, it's called somehow inside setuptools when running `python setup.py test`

> safest thing to do if you _cannot_ allow anything to pull in dependencies, is to manually download or build the packages you need and serve them on a local mirror of PyPI, then configure both `easy_install` and `pip` to use that as the index

We don't need any of that - installation of required packages is handled by the native packaging system (FreeBSD ports). However, we need consistent and safe behavior in cases where dependency lists are incomplete or outdated (which is expected when developing and updating packages) - python build should hard fail instead of fetching from the network in these cases.

**Comment by pganssle** [comment]
@AMDmi3 Presumably the safest way to prevent that would be to disable the network during the build, no? You can also null-route the indices by modifying the distutils and/or pip configuration.

I'm not saying that we *can't* fix this - and maybe there's already an option I don't know about that fixes it - but I still think your best bet for "don't let these things make network requests" is to actually lock down the code's ability to make make network requests.

**Comment by AMDmi3** [comment]
> @AMDmi3 Presumably the safest way to prevent that would be to disable the network during the build, no?

Yes, it's done on build servers when official packages are built, however not when users build their ports by hand. I'm not sure there's a way to do it from plain user on FreeBSD.

> You can also null-route the indices by modifying the distutils and/or pip configuration.

Umm, any way to do it without making any permanent changes to the filesystem outside of build directory? E.g. through environment vars directly, or by creating a temporary config and forcing it through environment var?

> is to actually lock down the code's ability to make make network requests

Yes, but I think that the free/open source software should be well behaved and customizable so there won't be a reason to limit it with other software.
